### PR TITLE
shiwei.fu

### PR DESCRIPTION
**Pull request by LogosFu**

### CODE REVIEW

**Comment by maguangguang** [comment]
Well:
1. GOOD: test coverage 100%
2. good baby steps commit  and readable message
3. GOOD: use LINQ
4. each commit contains test and implementation codes

Less Well:
1. some variables naming could be more meaningful

Suggestion:
1. add tests for the edge case like '4A0B', '0A4B'

File: BullsAndCows/BullsAndCowsGame.cs [hunk 1]
```diff
@@ -1,20 +1,58 @@
...
+using System.Linq;
 
 namespace BullsAndCows
 {
     public class BullsAndCowsGame
     {
-        private readonly SecretGenerator secretGenerator;
-        public BullsAndCowsGame(SecretGenerator secretGenerator)
+        private readonly ISecretGenerator secretGenerator;
+        private readonly string screit;
```
Comment (maguangguang) [inline BullsAndCows/BullsAndCowsGame.cs:10]: Typo: screit -> secret

File: BullsAndCows/BullsAndCowsGame.cs [hunk 2]
```diff
@@ -1,20 +1,58 @@
...
+            var inputNumberSet = inputNumbers.ToHashSet();
+            if (inputNumberSet.Count != inputNumbers.Length)
+            {
+                return "Wrong Input, input again";
+            }
+
+            if (inputNumbers.Length != 4)
+            {
+                return "Wrong Input, input again";
+            }
```
Comment (maguangguang) [inline BullsAndCows/BullsAndCowsGame.cs:35]: It's better to extract a method

File: BullsAndCows/BullsAndCowsGame.cs [hunk 3]
```diff
@@ -1,20 +1,58 @@
...
+
+            var match = secret.Select((t1, i) => inputNumbers.Where((t, j) => t == t1 && i == j).Count()).Sum();
+            var positionErrorMatch = secret.Select((t1, i) => inputNumbers.Where((t, j) => t == t1).Count()).Sum();
```
Comment (maguangguang) [inline BullsAndCows/BullsAndCowsGame.cs:38]: Naming: 'positionErrorMatch' not only contains error matched digit, also contains correct matched digit
it's better to rename like:
match -> valueAndPositionMatchedCount
positionErrorMatch -> valueMatchedCount

File: BullsAndCows/SecretGenerator.cs [hunk 4]
```diff
@@ -2,11 +2,22 @@
...
 namespace BullsAndCows
 {
-    public class SecretGenerator
+    public class SecretGenerator : ISecretGenerator
     {
-        public virtual string GenerateSecret()
+        public string GenerateSecret()
         {
-            throw new NotImplementedException();
+            var str = string.Empty;
```
Comment (maguangguang) [inline BullsAndCows/SecretGenerator.cs:9]: Naming: str -> secret

File: BullsAndCowsTest/BullsAndCowsGameTest.cs [hunk 5]
```diff
@@ -11,7 +13,131 @@ public void Should_create_BullsAndCowsGame()
...
-            Assert.True(game.CanContinue);
+            Assert.True(game.IsCanContinue());
+        }
+
+        [Fact]
+        public void Should_generate_secret_when_game_init()
+        {
+            //given
+            var mockGenerator = new Mock<ISecretGenerator>();
+            const string secret = "Secret";
```
Comment (maguangguang) [inline BullsAndCowsTest/BullsAndCowsGameTest.cs:24]: Use a value which is fulfill the requirement, otherwise it is a little confused

File: BullsAndCows/BullsAndCowsGame.cs [hunk 6]
```diff
@@ -1,20 +1,58 @@
...
+            count += 1;
+            if (count >= 6)
```
Comment (maguangguang) [inline BullsAndCows/BullsAndCowsGame.cs:40]: magic codes, define a const variable like MAX_GUESS_TIMES = 6
